### PR TITLE
inference: make getfield_tfunc more robust for abstract PartialStruct

### PR DESCRIPTION
**Pull request by vtjnash**
We're currently careful never to make these. But good to be careful?

refs #34513

### CODE REVIEW

File: base/compiler/tfuncs.jl [hunk 1]
```diff
@@ -727,17 +727,21 @@ function getfield_tfunc(@nospecialize(s00), @nospecialize(name))
             end
         end
         s = typeof(sv)
-    elseif isa(s, PartialStruct)
+    elseif isa(s00, PartialStruct)
+        s01 = widenconst(s00)
+        s = unwrap_unionall(s01)::DataType
```
Comment (vtjnash) [inline base/compiler/tfuncs.jl]:
```suggestion
        s = s01::DataType
```
currently this simplification would hold, but do we want to assume that?

**Comment by vtjnash** [comment]
It turns out we might make abstract PartialStruct for Tuple (refs https://github.com/JuliaLang/julia/pull/39402), so reopening to add vararg handling to this also

**Comment by martinholters** [comment]
So is it possible to construct a case that fails with the current code?

**Review by aviatesk — APPROVED** [review]
LGTM. AFAIU, `Vararg` handling was already added https://github.com/JuliaLang/julia/pull/37769.
Still needs rebase though.

File: base/compiler/tfuncs.jl [hunk 2]
```diff
@@ -826,17 +826,18 @@ function getfield_tfunc(@nospecialize(s00), @nospecialize(name))
             end
         end
         s = typeof(sv)
-    elseif isa(s, PartialStruct)
+    elseif isa(s00, PartialStruct)
+        s = widenconst(s00)
+        sty = unwrap_unionall(s)::DataType
```
Comment (aviatesk) [inline base/compiler/tfuncs.jl:831]: Just to make sure: is there any possibility where `s::DataType` doesn't hold ?
This check is good to have though, we may want to form `PartialStruct` for abstract types other than tuples in the future.

**Review by aviatesk — APPROVED** [review]

File: base/compiler/tfuncs.jl (hunk 2 above)
Comment (vtjnash) [reply base/compiler/tfuncs.jl:831]: Currently, no. That is why this PR has sat for so long untouched. But turns out to be fairly straightforward here to add (though I think the necessary type lattice changes may currently be difficult).

**Pull request merged**